File: unn/graphs/makeDataForClassificationExperiment.R
data_path <- "D:/Dev/bio/data"
setwd(data_path)

files <- dir(".", pattern="csv")
#files <- c("geneMeanMatsTCGABRCA5.rd")
file <- files[1]

#for (file in files) {
  setwd(data_path)

  cancerType <- sub(".csv", "", file)
  
  print(paste("Process:", cancerType))
  
  data <- read.csv(file, header=F, sep=",", dec=".")
  phen <- data["V1"] == "cancer"
  
  numCancer <- sum(phen)
  numHealthy <- sum(!phen)
  minExamples <- min(numCancer, numHealthy)
  
  indexes <- seq_along(phen)
  cancerIndexes <- indexes[phen]
  healthyIndexes <- indexes[!phen]
  
  
  if (minExamples < 50) {
    numChunks <- 2
  } else {
    numChunks <- 4
  }
  
  # shuffle indexes
  cancerIndexes <- sample(cancerIndexes)
  healthyIndexes <- sample(healthyIndexes)
  
  cancerChunks <- split(1:numCancer, cut(1:numCancer, numChunks))
  healthyChunks <- split(1:numHealthy, cut(1:numHealthy, numChunks))
  
  dir.create(file.path(data_path, cancerType))
  setwd(file.path(data_path, cancerType))
  
  for (i in 1:numChunks) {
    testCancer <- cancerIndexes[cancerChunks[[i]]]
    trainCancer <- cancerIndexes[!(cancerIndexes %in% testCancer)]
    testHealthy <- healthyIndexes[healthyChunks[[i]]]
    trainHealthy <- healthyIndexes[!(healthyIndexes %in% testHealthy)]
    
    write.table(data[trainCancer, ], file=paste(cancerType, "_train_cancer_", i, ".csv", sep=""),
                row.names=F, col.names=F, quote=F, sep=",")
    write.table(data[trainHealthy, ], file=paste(cancerType, "_train_healthy_", i, ".csv", sep=""),
                row.names=F, col.names=F, quote=F, sep=",")
    
    testIndexes <- c(testCancer, testHealthy)
    write.table(data[testIndexes, ], file=paste(cancerType, "_test_", i, ".csv", sep=""),
                row.names=F, col.names=F, quote=F, sep=",")
    
    testLabels <- c(rep("cancer", length(testCancer)), rep("healthy", length(testHealthy)))
    write.table(testLabels, file=paste(cancerType, "_test_labels", i, ".csv", sep=""),
                row.names=F, col.names=F, quote=F, sep=",")
  }
#}

File: unn/graphs/makeDataForClassificationExperiment.py
import sys
from itertools import islice
from os import listdir, mkdir
from os.path import isfile, join
from random import shuffle

def chunker(seq, size):
    return (seq[pos:pos + size] for pos in xrange(0, len(seq), size))

data_path = sys.argv[1]
files = [ f for f in listdir(data_path) if isfile(join(data_path,f)) ]

f = files[0]

cancerIndexes = []
healthyIndexes = []
with open(join(data_path,f)) as file:
    index = 0
    while True:
        next_line = list(islice(file, 1))
        if not next_line:
            break
        
        if next_line[0].startswith("cancer"):
            cancerIndexes.append(index)
        else:
            healthyIndexes.append(index)
        index = index + 1

if len(healthyIndexes) < 50:
    numChunks = 2
else:
    numChunks = 4

shuffle(healthyIndexes)
print healthyIndexes
numHealthy = len(healthyIndexes)
chunks = list(chunker(healthyIndexes, numHealthy/numChunks))

work_dir = join(data_path, f[0:4])
mkdir(work_dir)

for chunkIdx in range(numChunks):
    output_dir = join(work_dir, "test_" + str(chunkIdx))
    mkdir(output_dir)
    
    train_file = open(join(output_dir, "train.csv"), "w")
    test_file = open(join(output_dir, "test.csv"), "w")
    test_labels_file = open(join(output_dir, "test_labels.csv"), "w")
    
    with open(join(data_path,f)) as file:
        index = 0
        while True:
            next_line = list(islice(file, 1))
            if not next_line:
                break
            data = next_line[0].split(", ")
            if (index in cancerIndexes) or not (index in chunks[chunkIdx]):
                test_file.write(", ".join(data[1:]))
                if (index in cancerIndexes):
                    test_labels_file.write("cancer\n")
                else:
                    test_labels_file.write("healthy\n")
            else:
                train_file.write(", ".join(data[1:]))
            
            index = index + 1
    
    train_file.close()
    test_file.close()
    test_labels_file.close()
    
    
print cancerIndexes
print healthyIndexes
